# HONOUR CODE:

I solemnly swear that I have not discussed my assignment solutions with anyone in any way and the solutions I am submitting are my own personal work.

Full Name: [Pranamya Prakash Korde]

# Assignment 3 Machine Learning.


## Question 1.

In this question we will use `US Census Income Dataset` and will use different variations of NB (Naive Bayes) classifier to predict the `annual_income` target feature. 


### Part A: Data Preparation.

In this Part we are going to perform 2 specific tasks.

In Task 1: we are going to perform equal-width binning on the numeric features `age` and `education_years` into 3 bins namely 'low', 'mid', 'high'. and we save this into the data frame named `df_all_cat`.

In Task 2 : in this task, we are going to perform One Hot Encoding on the dataframe `df_all_cat` and then save the encoded data into new data frame  `df_all_cat_ohe`.

For this purpose, we read the dataset into python 3.0 environment and we will perform basic exploration tasks like:

- checking shape and data type of each of the column.
- removing ID like columns from the data set.

In [1]:
import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np
import altair as alt

pd.set_option('display.max_columns', None) 


dataset = pd.read_csv("A3_Q1_train.csv")
dataset.head()

row_id  age  education_years    workclass      marital_status  \
0       1   48               14    Local-gov            Divorced   
1       2   23               13    Local-gov       Never-married   
2       3   45               13    Local-gov       Never-married   
3       4   51               13  Federal-gov  Married-civ-spouse   
4       5   51               14    Local-gov  Married-civ-spouse   

        occupation annual_income  
0   Prof-specialty   high_income  
1   Prof-specialty    low_income  
2   Prof-specialty    low_income  
3  Exec-managerial    low_income  
4   Prof-specialty   high_income

In [2]:
dataset.shape

(500, 7)

In [3]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   row_id           500 non-null    int64 
 1   age              500 non-null    int64 
 2   education_years  500 non-null    int64 
 3   workclass        500 non-null    object
 4   marital_status   500 non-null    object
 5   occupation       500 non-null    object
 6   annual_income    500 non-null    object
dtypes: int64(3), object(4)
memory usage: 27.5+ KB


In [4]:
dataset = dataset.drop(columns = 'row_id')
dataset.head()

age  education_years    workclass      marital_status       occupation  \
0   48               14    Local-gov            Divorced   Prof-specialty   
1   23               13    Local-gov       Never-married   Prof-specialty   
2   45               13    Local-gov       Never-married   Prof-specialty   
3   51               13  Federal-gov  Married-civ-spouse  Exec-managerial   
4   51               14    Local-gov  Married-civ-spouse   Prof-specialty   

  annual_income  
0   high_income  
1    low_income  
2    low_income  
3    low_income  
4   high_income

### Task 1 

In this task we will perform equal width binning on numeric features `age` and `education_years` and save the dataset into new data frame called `df_all_cat`.

In [5]:
dataset['age'] = pd.cut(dataset['age'], 3, labels = ["low", "mid", "high"])
dataset['education_years'] = pd.cut(dataset['education_years'], 3, labels = ["low", "mid", "high"])
df_all_cat = dataset
df_all_cat.head()

age education_years    workclass      marital_status       occupation  \
0  high            high    Local-gov            Divorced   Prof-specialty   
1   low            high    Local-gov       Never-married   Prof-specialty   
2   mid            high    Local-gov       Never-married   Prof-specialty   
3  high            high  Federal-gov  Married-civ-spouse  Exec-managerial   
4  high            high    Local-gov  Married-civ-spouse   Prof-specialty   

  annual_income  
0   high_income  
1    low_income  
2    low_income  
3    low_income  
4   high_income

In [6]:
for col in df_all_cat.columns.to_list():  
    print(col + ':')
    print(df_all_cat[col].value_counts())
    print('******************************')

age:
mid     230
low     140
high    130
Name: age, dtype: int64
******************************
education_years:
high    304
mid     193
low       3
Name: education_years, dtype: int64
******************************
workclass:
Local-gov      225
State-gov      148
Federal-gov    127
Name: workclass, dtype: int64
******************************
marital_status:
Married-civ-spouse    230
Never-married         155
Divorced              115
Name: marital_status, dtype: int64
******************************
occupation:
Prof-specialty     224
Adm-clerical       159
Exec-managerial    117
Name: occupation, dtype: int64
******************************
annual_income:
low_income     320
high_income    180
Name: annual_income, dtype: int64
******************************


### Task 2 

In this task, we are going to perform One hot Encoding on the data set `df_all_cat` to perform binary encoding. we will use `get_dummies()`function for binary encoding. Due to the binary encoding features with 2 unique values are encoded into a signle feature while features with more than 2 unique values are encoded into (n-1) unique features where 'n' represents the unique levels of the original feature column.

For this purpose we will store all the feature names into a single list and then perform One hot encoding on every feature in the list.

In [7]:
categorical_cols = df_all_cat.columns.tolist()

for col in categorical_cols:
    n = len(df_all_cat[col].unique())
    if (n == 2):
        df_all_cat[col] = pd.get_dummies(df_all_cat[col], drop_first=True)

df_all_cat = pd.get_dummies(df_all_cat)
df_all_cat_ohe = df_all_cat

we can see from the encoded data set that the target feature in encoded `1 : low_income` and `0 : high_income`. But as stated in the requirement the postive class is 1 so we need to encode the target variable accordingly.

In [8]:
df_all_cat_ohe['annual_income'] = np.where(df_all_cat_ohe['annual_income'] == 0 , 1 , 0) 

In [9]:
print(df_all_cat_ohe.shape)

(500, 16)


In [10]:
df_all_cat_ohe.head()

annual_income  age_low  age_mid  age_high  education_years_low  \
0              1        0        0         1                    0   
1              0        1        0         0                    0   
2              0        0        1         0                    0   
3              0        0        0         1                    0   
4              1        0        0         1                    0   

   education_years_mid  education_years_high  workclass_Federal-gov  \
0                    0                     1                      0   
1                    0                     1                      0   
2                    0                     1                      0   
3                    0                     1                      1   
4                    0                     1                      0   

   workclass_Local-gov  workclass_State-gov  marital_status_Divorced  \
0                    1                    0                        1   
1                    1                    0                        0   
2                    1                    0                        0   
3                    0                    0                        0   
4                    1                    0                        0   

   marital_status_Married-civ-spouse  marital_status_Never-married  \
0                                  0                             0   
1                                  0                             1   
2                                  0                             1   
3                                  1                             0   
4                                  1                             0   

   occupation_Adm-clerical  occupation_Exec-managerial  \
0                        0                           0   
1                        0                           0   
2                        0                           0   
3                        0                           1   
4                        0                           0   

   occupation_Prof-specialty  
0                          1  
1                          1  
2                          1  
3                          0  
4                          1

## Part  B : Bernoulli NB


Bernoulli Naive Bayes assumes that all of the descriptive features are binary and have a value of 0 or 1. In this part, we will apply Bernoulli NB to the One hot encoded (OHE) data with default hyperparameter values and check the accuracy score of the classifier on our data.

For that reason, we will separate the `target` feature from rest of the data set and store that dataset into `data` dataframe.

In [11]:
data = df_all_cat_ohe.drop(columns = 'annual_income')
target = df_all_cat_ohe['annual_income']
print(data.shape)
print(target.shape)

(500, 15)
(500,)


In [12]:
from sklearn.naive_bayes import BernoulliNB

nb_classifier = BernoulliNB()
nb_classifier.fit(data,target)
Bern_def_score = nb_classifier.score(data,target)
Bern_def_score

0.83

The Accuracy score of a `Bernoulli NB with default parameters` on the given data is `0.83`. Now lets check the Accuracy score of the Gaussian Naive Bayes on the same data. As we all know, Gaussain NB perform good against Numerical data and assumes each descriptive feature followds Gaussian Probability distribution. But below we are applying the Gaussian NB on the binary discriptive features.


## Part C : Gaussian NB


Gaussian Naive Bayes assume all desciptive features follows Gaussian probability distribution. But that is not the case here, in our problem all the descriptive features are binary features having values 0 and 1. lets fit this algorithm on the data and check what result we get..

In [13]:
from sklearn.naive_bayes import GaussianNB

nb_classifier = GaussianNB()
nb_classifier.fit(data, target)
Gauss_def_score = nb_classifier.score(data, target)
Gauss_def_score

0.728

We can see from the above that, Gaussian Naive Bayes performed with accurcy score of `0.728` on the given data with default parameters. In the next task we are going to tune these above models to squeeze some additional performance from them.

## Part D : Tuning the Models


First to tune the Gaussian Naive Bayes model, we will vary the hyperparamter value `var_smoothing` in range ($10^{-1}$ to  $10^{-9}$) and we will take 80 values from the logspace. We will simply use for loop to calculate score for each var_smoothing value and store it into a dataframe df. After that we will plot the score values against the var_smoothing values to check what is the highest performance we achieved by tunning.

### Tuning Gaussian Naive Bayes:

In [14]:
df = pd.DataFrame(columns=['var_smoothing', 'score'])
for var in np.logspace(-1,-9, num=80):
    nb_classifier = GaussianNB(var_smoothing = var)
    nb_classifier.fit(data, target)
    df.loc[len(df)] = [var,nb_classifier.score(data,target)]

Gauss_tun_score = df['score'].max()
df.sort_values(by='score', ascending=False).head() 

var_smoothing  score
9       0.012263  0.832
6       0.024683  0.830
1       0.079202  0.830
8       0.015484  0.830
7       0.019550  0.830

In [15]:
alt.Chart(df, 
          title='Gaussian NB Performance Comparison'
         ).mark_line(point=True).encode(
    alt.X('var_smoothing', title='Var. Smoothing'),
    alt.Y('score', title='Scores', scale=alt.Scale(zero=False))
).interactive()

alt.Chart(...)

We can see from the above graph that score value increases with the var_smoothing score and achieve maximum score of approximately `0.832` at `var_smoothing 0.01226`.

In the next step we try to tune the Bernoulli NB model on the given data by varying the hyperparameter `alpha` value. When we dont want to consider alpha its value is 0 by default. we are varying it in range of (1 to 200 by increment of 5) i.e. 40 values. We will check the score of the maximum performance achieved by this model.

In [16]:
df = pd.DataFrame(columns=['alpha', 'score'])
for alpha in np.arange(1,200, step = 5):
    nb_classifier = BernoulliNB(alpha = alpha)
    nb_classifier.fit(data,target)
    df.loc[len(df)] = [alpha,nb_classifier.score(data,target)]

Bern_tun_score = df['score'].max()
df.sort_values(by = "score", ascending = False).head()

alpha  score
27  136.0  0.842
28  141.0  0.842
29  146.0  0.842
35  176.0  0.840
34  171.0  0.840

In [17]:
alt.Chart(df, 
          title='Bernoulli NB Performance Comparison'
         ).mark_line(point=True).encode(
    alt.X('alpha', title='alpha values'),
    alt.Y('score', title='Scores', scale=alt.Scale(zero=False))
).interactive()

alt.Chart(...)

We can see from the above results that, even after HP tuning, the performance of Bernoulli NB does not changed by great difference. The new optimal performance is `0.842` at `alpha equal 136`.

## Part E : Hybrid NB

As we can understand from above that, at default parameters, Gaussian naive bayes will not perform that good as compared to Bernoulli NB when the data set contains binary descriptive features. In real world cases, most of the dataset contains desciptive features of both the types numeric as well as categorical.
To achieve the optimal performance from both kind of features, we can use the combination of Gaussian as well as Bernoulli NB classfiers where, Gaussian NB can be used on the Numeric features and Bernoulli NB can be applied to categorical features.

We can do this because, while applying any NB technique we assume the naive assumption that, all the descriptive features are independent from each other when we consider a target column as a third dependent feature (partial independence).

In [18]:
df = pd.read_csv("A3_Q1_train.csv")
df = df.drop(columns = "row_id")

In [19]:
df.head()

age  education_years    workclass      marital_status       occupation  \
0   48               14    Local-gov            Divorced   Prof-specialty   
1   23               13    Local-gov       Never-married   Prof-specialty   
2   45               13    Local-gov       Never-married   Prof-specialty   
3   51               13  Federal-gov  Married-civ-spouse  Exec-managerial   
4   51               14    Local-gov  Married-civ-spouse   Prof-specialty   

  annual_income  
0   high_income  
1    low_income  
2    low_income  
3    low_income  
4   high_income

Hence , we are splitting the data frame into 2 data frames, df_Num which is numerical data frame and df_col is categorical dataframe. we will then apply Gaussian NB on the first one and Bernoulli NB on the later one. we will check their performance by checking the accuracy score of each of the individual. We are using default parameters in each of the model fitting process.


### 1. Applying Gaussian NB on numeric data

In [20]:
target = df['annual_income']
target = target.replace({"high_income": 1, "low_income": 0})
df_num = df.loc[:, {"age", "education_years"}]

In [21]:
nb_classifier = GaussianNB()
nb_classifier.fit(df_num, target)
Gauss_score = nb_classifier.score(df_num, target)
Gauss_score

0.72

We can see that the performance of Gaussian NB is `0.726` on the numeric featured dataset which is higher than that on the whole dataset when default hyperparameter values are considered.

### 2. Applying Benoulli NB on the binary featured data.

For this step first we have to encode the categorical features with the binary encoding technique One Hot Encoding.

In [22]:
df = df.drop(columns = 'annual_income')
df_col = pd.DataFrame()
for col in df.columns[df.dtypes==object].tolist():
    df_col[col] = df[col]

In [23]:
cat_cols = df_col.columns.tolist()

for col in cat_cols:
    n = len(df_col[col].unique())
    if (n == 2):
        df_col[col] = pd.get_dummies(df_col[col], drop_first=True)

df_col = pd.get_dummies(df_col)
df_col.head()

workclass_Federal-gov  workclass_Local-gov  workclass_State-gov  \
0                      0                    1                    0   
1                      0                    1                    0   
2                      0                    1                    0   
3                      1                    0                    0   
4                      0                    1                    0   

   marital_status_Divorced  marital_status_Married-civ-spouse  \
0                        1                                  0   
1                        0                                  0   
2                        0                                  0   
3                        0                                  1   
4                        0                                  1   

   marital_status_Never-married  occupation_Adm-clerical  \
0                             0                        0   
1                             1                        0   
2                             1                        0   
3                             0                        0   
4                             0                        0   

   occupation_Exec-managerial  occupation_Prof-specialty  
0                           0                          1  
1                           0                          1  
2                           0                          1  
3                           1                          0  
4                           0                          1

In [24]:
nb_classifier = BernoulliNB()
nb_classifier.fit(df_col,target)
Bern_Score = nb_classifier.score(df_col,target)
Bern_Score

0.816

As we can see the performance of Bernoulli NB on the rest of the data set after binary encoding is performed is `0.816` when we consider the default hyperparameter values.

**The Naive Bayes Classifier naively assumes that, each of the discriptive feature in the domain is conditionally independent of all of the other discriptive features, given the state of the target feature.**

Hence, simply taking the average of both of the above performances can give us the resultant performance value of the combined NB (Gaussian and Bernoulli) or Hybrid NB. 

In [25]:
Hybrid_score = (Gauss_score+Bern_Score)/2
Hybrid_score

0.768

## Part F : Wrapping Up


In this part we are going to wrap up the assignment  done above by storing the methods and their corresponding accuracy scores into the dataset named `df_summary`.

In [26]:
df_summary = pd.DataFrame(columns = ['method', 'Accuracy'])
df_summary.loc[len(df_summary)] = ['Bernoulli NB Default', Bern_def_score]
df_summary.loc[len(df_summary)] = ['Gaussian NB Default', Gauss_def_score]
df_summary.loc[len(df_summary)] = ['Bernoulli NB Tuned', Bern_tun_score]
df_summary.loc[len(df_summary)] = ['Gaussian NB Tuned', Gauss_tun_score]
df_summary.loc[len(df_summary)] = ['Hybrid  NB Default', Hybrid_score]
df_summary

method  Accuracy
0  Bernoulli NB Default     0.830
1   Gaussian NB Default     0.728
2    Bernoulli NB Tuned     0.842
3     Gaussian NB Tuned     0.832
4    Hybrid  NB Default     0.768

1. Hyperparameter Tuning helps in the improving the performance of a classifier. In our case, we have tuned the hyperparameter `alpha` in case of Bernoulli NB and `var_smoothing` in case of Gaussian NB classifier. Accuracy score of both the classifiers increase by some amount. Bernoulli NB accuracy score increased from `0.830` to `0.842` after HP tuning which not that much. Gaussian NB accuracy score increased from `0.728` to `0.832` which is comparatively higher than increase of bernoulli NB performance.

2. Default Hybrid NB is implemented by spliting the dataset into numerical and categorical discriptive features and applying Gaussian NB and Bernoulli NB to the respective ones. We can see from the above score that, Accuracy score of Hybrid NB with default hyperparameters is `0.768` which is definitely higher than `untuned` Gaussian NB but is less than untuned Bernoulli NB by almost `0.06` score.